subworkflows/nf-core/dia_proteomics_analysis/main.nf

```diff
@@ -0,0 +1,426 @@
+
+    DIANN_INSILICOLIBRARYGENERATION(ch_insilico_library_input)
+    ch_speclib = ch_speclib.mix(DIANN_INSILICOLIBRARYGENERATION.out.predict_speclib)
+    ch_versions = ch_versions.mix(DIANN_INSILICOLIBRARYGENERATION.out.versions)
+
+    // In-silico libraries have been generated for combinations of FASTA and configuration, which 
+    // may have been the same over multiple inputs. Use a combine to annotate inputs with in silico libraries.
+    ch_fasta_input_with_speclib = ch_speclib
+        .map{ meta, speclib -> [meta.findAll { key, value -> key != 'config' }, speclib] } // Filter out config from meta to allow the combine by: 0
+        .combine(input.ms_file_fasta, by: 0)
+        .map { meta_fasta_enzyme, speclib, meta_input_fasta, ms_file ->
+            [meta_input_fasta, ms_file, speclib]
+        }
+    
+    //
+    // MODULE: Preliminary analysis
+    //
+
+    // We only need to do preliminary analysis and empirical library assembly if no empirical library is provided
+    // Conditional execution pattern:
+    //   - join with remainder: true preserves inputs even if no empirical library exists
+    //   - filter keeps only tuples containing null (i.e., no pre-generated empirical library)
+    //   - downstream processes only execute when needed
+
+    preliminary_branches = ch_fasta_input_with_speclib  
+        .join(ch_empirical, remainder: true)
+        .filter { tuple ->
+            tuple.any { it == null } // ch_empirical was an empty channel
+        }
+        .map { meta_input_fasta, ms_file, speclib, _ ->
+            [meta_input_fasta, ms_file, speclib]
+        }
+        .branch { 
+            random_preanalysis: random_preanalysis
+            no_random_preanalysis: no_random_preanalysis
+        }
+
+    ch_preliminary_analysis_input = preliminary_branches.no_random_preanalysis
+        .mix(
+            preliminary_branches.random_preanalysis
+                .toSortedList{ a, b -> file(a[1]).getName() <=> file(b[1]).getName() }
+                .flatMap()
+                .randomSample(random_preanalysis_n, random_preanalysis_seed)
+        )
+
+    DIANN_PRELIMINARYANALYSIS(
+        ch_preliminary_analysis_input.map { meta, ms_file, speclib ->
+            [meta, ms_file, [], [], speclib, []]  // DIANN module input: [meta, ms_files, ms_file_names, fasta, library, quant]
+        }
+    )
+    ch_versions = ch_versions.mix(DIANN_PRELIMINARYANALYSIS.out.versions)
+
+    //
+    // MODULE: Assemble empirical library with all inputs from the same experiment + search DB
+    //
+    ch_empirical_input = ch_preliminary_analysis_input    // [meta_input_fasta, ms_file, speclib]
+        .join(DIANN_PRELIMINARYANALYSIS.out.diann_quant)  // [meta_input_fasta, ms_file, speclib, diann_quant]
+        .map { [it[0].experiment] + it.drop(1)}           // [meta_exp_searchdb, ms_file, speclib, diann_quant]
+        .groupTuple()                                     // [meta_exp_searchdb, [ms_file], [speclib], [diann_quant]]
```

Comment:
Not sure if it is a concern but each experiment should only have one speclib, so there should only be a single entry in`[speclib]` .